src/Library/demos/View Switcher/main.py

```diff
@@ -0,0 +1,39 @@
+import gi
+
+gi.require_version("Gtk", "4.0")
+gi.require_version("Adw", "1")
+from gi.repository import Gtk, Adw
+import workbench
+
+notifications_page: Adw.ViewStackPage = workbench.builder.get_object("page3")
+notification_list: Gtk.ListBox = workbench.builder.get_object("notification_list")
+
+notification_count = 5
+notifications_page.set_badge_number(notification_count)
+
+
+def on_clicked(button):
+    notifications_page.set_badge_number(notifications_page.get_badge_number() - 1)
+    notification_list.remove(button.parent_row)
+
+    if not notifications_page.get_badge_number():
+        notifications_page.set_needs_attention(False)
+
+
+for i in range(notification_count):
+    notification_row = Adw.ActionRow(title="Notification", selectable=False)
+
+    button = Gtk.Button(
+        halign=Gtk.Align.CENTER,
+        valign=Gtk.Align.CENTER,
+        margin_top=10,
+        margin_bottom=10,
+        icon_name="check-plain-symbolic",
+    )
+
+    button.parent_row = notification_row
```

Comment:
I don't really like this. I wasn't even sure what this would be doing, seeing as this is a GObject and not a normal Python object, but it seems the Python representations for GObjects also have a __dict__ where this information gets stored. As far as I can tell this isn't even really documented and I'm also not sure if this info would still be guaranteed to be there if you got a reference to the GObject through somewhere else. It probably would, but I still feel like this isn't great, just adding arbitrary attributes to an object outside of it isn't considered good practice anyway.

I think the only way to solve this is wrapping everything in the loop in another function and then placing `on_clicked` inside of that function, effectively turning it into a closure for the specific function call.
Unless you have a better idea I think this would be the way to go for now.

Reply:
To clarify: In that scenario the loop would just contain one line that calls this new function. The function contains the old body of the loop and `on_clicked`.

Reply:
I suggest a different solution: Pass the `notification_row` as `user_data` to the `on_clicked` callback function, i.e.

```python
    notification_row = Adw.ActionRow(title="Notification", selectable=False)

    button = Gtk.Button(
        halign=Gtk.Align.CENTER,
        valign=Gtk.Align.CENTER,
        margin_top=10,
        margin_bottom=10,
        icon_name="check-plain-symbolic",
    )

    button.connect("clicked", on_clicked, notification_row)
```

and

```python
def on_clicked(button, notification_row):
    notifications_page.set_badge_number(notifications_page.get_badge_number() - 1)
    notification_list.remove(notification_row)

    if not notifications_page.get_badge_number():
        notifications_page.set_needs_attention(False)
```

Reply:
Otherwise this port looks good to me.